### PR TITLE
[TR] TIP-458: Move converters from processors to readers

### PR DESCRIPTION
**Pull request by pierallard**
# What!

Before, processors took data from readers in "flat" format, converted it to have standard format, then processed it.
Now, that's the readers' roles to convert data! So, processors only take standard format to process it :heart:

This implies some modifications:
- services have their names changed (processor.xxx.flat to processor.xxx)
- some readers moved from specific folders to "file format" folders (xlsx, csv, yaml)
- we must declare a specific reader per entity format (and no more aliases)
- the patapouille for products have moved from converter to processor, and is simplified (always ugly and product-specific, another card for this)

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | n
| Added Behats                      | n
| Changelog updated                 | y
| Review and 2 GTM                  | TODO
| Micro Demo to the PO (Story only) | n
| Migration script                  | y
| Tech Doc                          | n

### CODE REVIEW

File: src/Pim/Bundle/ConnectorBundle/Resources/config/processors.yml [hunk 1]
```diff
@@ -23,7 +23,6 @@ services:
     pim_connector.processor.denormalization.attribute_option.flat:
```
Comment (nidup) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/processors.yml]: Can become "pim_connector.processor.denormalization.attribute_option" (no more ".flat").
We can add a sed command to change the use of old service names

File: src/Pim/Component/Connector/Reader/File/Product/CsvProductReader.php [hunk 2]
```diff
@@ -45,4 +49,41 @@ public function read()
...
+        ];
+    }
+
+    /**
+     * @return array
+     */
+    protected function getMapping()
+    {
+        $jobParameters = $this->stepExecution->getJobParameters();
+        var_dump($jobParameters->all());
```
Comment (nidup) [inline src/Pim/Component/Connector/Reader/File/Product/CsvProductReader.php]: :fire:

File: src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml [hunk 3]
```diff
@@ -38,6 +38,13 @@ parameters:
 
 services:
     # array converters
+    pim_connector.array_converter.identity:
```
Comment (nidup) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml]: identity -> dummy?

File: src/Pim/Bundle/ConnectorBundle/Resources/config/batch_jobs/xlsx_connector.yml [hunk 4]
```diff
@@ -11,12 +11,12 @@ connector:
...
                     services:
                         reader:    pim_connector.reader.file.xlsx_product
-                        processor: pim_connector.processor.denormalization.product.flat
+                        processor: pim_connector.processor.denormalization.product
                         writer:    pim_connector.writer.doctrine.product
                 import_associations:
                     services:
-                        reader:    pim_connector.reader.file.xlsx_association_type
-                        processor: pim_connector.processor.denormalization.product_association.flat
+                        reader:    pim_connector.reader.file.xlsx_association
```
Comment (nidup) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/batch_jobs/xlsx_connector.yml]: xlsx_association_type or xlsx_association? you're fixing a previous typo maybe?

File: src/Pim/Bundle/ConnectorBundle/Resources/config/readers.yml [hunk 5]
```diff
@@ -3,15 +3,17 @@ parameters:
     pim_connector.reader.dummy_item.class:    Pim\Component\Connector\Reader\DummyItemReader
     pim_connector.reader.doctrine.base.class: Pim\Component\Connector\Reader\Doctrine\BaseReader
 
-    pim_connector.reader.file.csv.class:         Pim\Component\Connector\Reader\File\CsvReader
-    pim_connector.reader.file.csv_product.class: Pim\Component\Connector\Reader\File\Product\CsvProductReader
+    pim_connector.reader.file.csv.class:               Pim\Component\Connector\Reader\File\CsvReader
+    pim_connector.reader.file.csv_variant_group.class: Pim\Component\Connector\Reader\File\CsvVariantGroupReader
+    pim_connector.reader.file.csv_association.class:   Pim\Component\Connector\Reader\File\CsvAssociationReader
+    pim_connector.reader.file.csv_product.class:       Pim\Component\Connector\Reader\File\Product\CsvProductReader
```
Comment (nidup) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/readers.yml]: the classes location seems pretty inconsistent, there is also an issue with parameters, imho, CsvReader, CsvProductReader, CsvVariantGroupReader should be in the same folder, can be directly in Pim\Component\Connector\Reader\File or we could do a subfolder per format Pim\Component\Connector\Reader\File\Csv and Pim\Component\Connector\Reader\File\Xlsx anyway the parameter should represent the namespace.

File: src/Pim/Component/Connector/ArrayConverter/FlatToStandard/Product.php [hunk 6]
```diff
@@ -211,20 +210,6 @@ protected function mapFields(array $item, array $options)
 
     /**
      * @param array $mappedItem
-     * @param array $defaultValues
-     *
-     * @return array
-     */
-    protected function defineDefaultValues(array $mappedItem, array $defaultValues)
```
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/FlatToStandard/Product.php]: so I guess you handle it in the processor :)

File: src/Pim/Component/Connector/Processor/Denormalization/ProductAssociationProcessor.php [hunk 7]
```diff
@@ -71,8 +65,11 @@ public function __construct(
      */
     public function process($item)
     {
-        $convertedItem = $this->convertItemData($item);
-        $identifier    = $this->getIdentifier($convertedItem);
+        $item = array_merge(
```
Comment (nidup) [inline src/Pim/Component/Connector/Processor/Denormalization/ProductAssociationProcessor.php]: Means you can have a missing `associations` key when you convert an item?
In this case, I would avoid to do an array_merge to add an empty associations array, I would prefer a,
```
    if (!isset($item['associations']) {
              $this->skipItemWithMessage($item, 'The product does not have defined associations');
    }
```

File: src/Pim/Component/Connector/Processor/Denormalization/ProductProcessor.php [hunk 8]
```diff
@@ -143,17 +135,34 @@ public function process($item)
     }
 
     /**
+     * Sets default enabled status if not set.
+     *
+     * @param array $item
+     *
+     * @return array
+     */
+    protected function setEnabledStatus($item)
```
Comment (nidup) [inline src/Pim/Component/Connector/Processor/Denormalization/ProductProcessor.php]: setDefaultEnabledStatus is more explicit

File: src/Pim/Component/Connector/Processor/Denormalization/ProductProcessor.php [hunk 9]
```diff
@@ -143,17 +135,34 @@ public function process($item)
...
+    {
+        $this->itemHasStatus = isset($item['enabled']);
```
Comment (nidup) [inline src/Pim/Component/Connector/Processor/Denormalization/ProductProcessor.php]: We should never do this kind of stuff `$this->itemHasStatus`, here we define an instance variable to store a variable only used in the scope of this method (i know that you just moved code but please be a boy scout and fix this :) ).
The following snippet will do the job.
```
if (!isset($item['enabled'])) {
       $item['enabled'] = $jobParameters = $this->stepExecution->getJobParameters()->get('enabled');
}
```

File: src/Pim/Bundle/ConnectorBundle/Resources/config/batch_jobs/xlsx_connector.yml (hunk 4 above)
Comment (pierallard) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/batch_jobs/xlsx_connector.yml]: Yes, it's a fix. We are in the "import_assocations", but we used associatoin_types reader :)

File: src/Pim/Component/Connector/ArrayConverter/FlatToStandard/Product.php (hunk 6 above)
Comment (pierallard) [inline src/Pim/Component/Connector/ArrayConverter/FlatToStandard/Product.php]: Of course ; it's not the converter's role to do this.

File: src/Pim/Component/Connector/Processor/Denormalization/ProductProcessor.php (hunk 9 above)
Comment (pierallard) [inline src/Pim/Component/Connector/Processor/Denormalization/ProductProcessor.php]: OK. The thing is that we need to keep if status comes from default value or not (to filter products to update). But you're right, this should not be kept in a class variable, it's in the same function.

File: src/Pim/Bundle/ConnectorBundle/Resources/config/batch_jobs/xlsx_connector.yml (hunk 4 above)
Comment (nidup) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/batch_jobs/xlsx_connector.yml]: ok!

**Comment by nidup** [comment]
GTM once comments fixed

File: src/Pim/Bundle/BaseConnectorBundle/Resources/config/processors.yml [hunk 10]
```diff
@@ -22,15 +22,13 @@ services:
             - '@pim_catalog.builder.product'
             - ['pim_catalog_file', 'pim_catalog_image']
 
-    pim_base_connector.processor.job_instance: '@pim_connector.processor.denormalization.job_instance.flat'
-
     # Normalization
-    # @deprecated will be removed in 1.7 use @pim_connector.processor.normalization.group.flat
+    # @deprecated will be removed in 1.7 use @pim_connector.processor.normalization.group
```
Comment (nidup) [inline src/Pim/Bundle/BaseConnectorBundle/Resources/config/processors.yml]: If not used you can drop these services and related classes (we target the removing of base connector bundle for the 1.6 release)

Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Resources/config/processors.yml]: @nidup They are used in vendor/akeneo/pim-community-dev/features/Context/fixtures/base_connector_deprecated_batch_jobs.yml ... Don't know what to do

Comment (nidup) [inline src/Pim/Bundle/BaseConnectorBundle/Resources/config/processors.yml]: The deprecated batch jobs file is/was only used by the builds which test the deprecated jobs on the CI (old imports) these jobs are imho not runnable anymore and should be totally droped, we maybe forget to drop this file.

File: src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml (hunk 3 above)
Comment (jjanvier) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml]: dummy please
we have no idea of what `identity` means at first sight
and `dummy` is widely in the PHP/Symfony world

File: src/Pim/Bundle/ConnectorBundle/Resources/config/readers.yml [hunk 11]
```diff
@@ -4,15 +4,19 @@ parameters:
     pim_connector.reader.database.product.class: Pim\Component\Connector\Reader\Database\ProductReader
     pim_connector.reader.database.base.class:    Pim\Component\Connector\Reader\Database\BaseReader
 
-    pim_connector.reader.file.csv.class:         Pim\Component\Connector\Reader\File\CsvReader
-    pim_connector.reader.file.csv_product.class: Pim\Component\Connector\Reader\File\Product\CsvProductReader
+    pim_connector.reader.file.csv.class:               Pim\Component\Connector\Reader\File\Csv\CsvReader
```
Comment (skeleton) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/readers.yml]: IMHO, as we add a folder `CSV`, we can also rename `CSVReader` to `Reader` (same for others services)

File: src/Pim/Component/Connector/Reader/File/Csv/CsvAssociationReader.php [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+namespace Pim\Component\Connector\Reader\File\Csv;
+
+/**
+ * Product Association CSV reader
+ *
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class CsvAssociationReader extends CsvReader
```
Comment (skeleton) [inline src/Pim/Component/Connector/Reader/File/Csv/CsvAssociationReader.php]: As your comment is `Product Association CSV reader`, this class should be named `CsvProductAssociationReader`

File: src/Pim/Component/Connector/Reader/File/Csv/CsvProductReader.php [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+use Pim\Component\Connector\Reader\File\FileIteratorFactory;
+use Pim\Component\Connector\Reader\File\MediaPathTransformer;
+
+/**
+ * Product csv reader
+ *
+ * This specialized csv reader exists to replace relative media path to absolute path, in order for later process to
+ * know where to find the files.
+ *
+ * @author    Gildas Quemener <gildas@akeneo.com>
```
Comment (skeleton) [inline src/Pim/Component/Connector/Reader/File/Csv/CsvProductReader.php]: ? maybe you've just move this file ?

File: src/Pim/Component/Connector/Reader/File/Csv/CsvVariantGroupReader.php [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    public function read()
+    {
+        $data = parent::read();
+
+        if (!is_array($data)) {
+            return $data;
+        }
+
+        if (isset($data['values'])) {
+            $data['values'] = $this->mediaPathTransformer->transform($data['values'], $this->fileIterator->getDirectoryPath());
```
Comment (damien-carcel) [inline src/Pim/Component/Connector/Reader/File/Csv/CsvVariantGroupReader.php]: Too long for one line

File: src/Pim/Component/Connector/Reader/File/Csv/CsvProductReader.php [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+        return [
+            'mapping'           => $this->getMapping(),
+            'with_associations' => false
+        ];
+    }
+
+    /**
+     * @return array
+     */
+    protected function getMapping()
```
Comment (skeleton) [inline src/Pim/Component/Connector/Reader/File/Csv/CsvProductReader.php]: do you really need to create an other method for that ?

File: src/Pim/Bundle/ConnectorBundle/Resources/config/readers.yml [hunk 16]
```diff
@@ -4,15 +4,19 @@ parameters:
...
-    pim_connector.reader.file.xlsx_product.class: Pim\Component\Connector\Reader\File\Product\XlsxProductReader
+    pim_connector.reader.file.xlsx.class:         Pim\Component\Connector\Reader\File\Xlsx\XlsxReader
+    pim_connector.reader.file.xlsx_product.class: Pim\Component\Connector\Reader\File\Xlsx\XlsxProductReader
+
+    pim_connector.reader.file.yaml.class: Pim\Component\Connector\Reader\File\Yaml\YamlReader
 
     pim_connector.reader.file.file_iterator.class:          Pim\Component\Connector\Reader\File\FileIterator
     pim_connector.reader.file.file_iterator_factory.class:  Pim\Component\Connector\Reader\File\FileIteratorFactory
-    pim_connector.reader.file.media_path_transformer.class: Pim\Component\Connector\Reader\File\Product\MediaPathTransformer
+    pim_connector.reader.file.media_path_transformer.class: Pim\Component\Connector\Reader\File\MediaPathTransformer
```
Comment (jjanvier) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/readers.yml]: is it normal you changed it?

**Comment by damien-carcel** [comment]
One tiny comment, then GTM.

File: src/Pim/Component/Connector/Reader/File/Csv/CsvAssociationReader.php [hunk 17]
```diff
@@ -0,0 +1,21 @@
+<?php
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Reader/File/Csv/CsvAssociationReader.php]: why is it all green?
if you just moved it, it shouldn't
and I wouldn't like to merge like it, we lost the file history

File: src/Pim/Component/Connector/Reader/File/Csv/CsvProductReader.php [hunk 18]
```diff
@@ -0,0 +1,77 @@
+<?php
+
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Reader/File/Csv/CsvProductReader.php]: same here about history

File: src/Pim/Component/Connector/Reader/File/Csv/CsvVariantGroupReader.php [hunk 19]
```diff
@@ -0,0 +1,53 @@
+<?php
+
+namespace Pim\Component\Connector\Reader\File\Csv;
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Reader/File/Csv/CsvVariantGroupReader.php]: same comment about file history

File: src/Pim/Component/Connector/Reader/File/MediaPathTransformer.php [hunk 20]
```diff
@@ -0,0 +1,71 @@
+<?php
+
+namespace Pim\Component\Connector\Reader\File;
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Reader/File/MediaPathTransformer.php]: same comment about history

**Comment by jjanvier** [comment]
really good :+1: 
I'd just like to talk with you @nidup IRL about the standard format before giving my GTM :)

File: src/Pim/Component/Connector/ArrayConverter/FlatToStandard/Product.php (hunk 6 above)
Comment (jjanvier) [inline src/Pim/Component/Connector/ArrayConverter/FlatToStandard/Product.php]: maybe we should talk about that just to be sure

**Comment by jjanvier** [comment]
please don't merge as long as you didn't fix the history problem

File: CHANGELOG-1.6.md [hunk 21]
```diff
@@ -342,3 +333,22 @@
 - Move `Pim\Component\Connector\ArrayConverter\Flat\Product\ColumnsMapper` to `Pim\Component\Connector\ArrayConverter\FlatToStandard\Product\ColumnsMapper`
 - Move `Pim\Component\Connector\ArrayConverter\Flat\Product\AttributeColumnInfoExtractor` to `Pim\Component\Connector\ArrayConverter\FlatToStandard\Product\AttributeColumnInfoExtractor`
 - Move `Pim\Component\Connector\ArrayConverter\Flat\Product\FieldSplitter` to `Pim\Component\Connector\ArrayConverter\FlatToStandard\Product\FieldSplitter`
+- Move `Pim\Component\Connector\Reader\File\Product\CsvReader` to `Pim\Component\Connector\Reader\File\Csv\CsvReader`. Change constructor to add `Pim\Component\Connector\Reader\File\FileIteratorFactory` and `Pim\Component\Connector\ArrayConverter\ArrayConverterInterface`.
```
Comment (skeleton) [inline CHANGELOG-1.6.md]: You've just added `Pim\Component\Connector\ArrayConverter\ArrayConverterInterface` no ? FileIteratorFactory was already in CsvReader

File: CHANGELOG-1.6.md [hunk 22]
```diff
@@ -342,3 +333,22 @@
...
+- Move `Pim\Component\Connector\Reader\File\CsvProductReader` to `Pim\Component\Connector\Reader\File\Csv\CsvProductReader`. Change constructor to remove `Pim\Component\Catalog\Repository\AttributeRepositoryInterface` and add `Pim\Component\Connector\Reader\File\FileIteratorFactory`, `Pim\Component\Connector\Reader\File\MediaPathTransformer` and `Pim\Component\Connector\ArrayConverter\ArrayConverterInterface`.
```
Comment (skeleton) [inline CHANGELOG-1.6.md]: same

File: src/Pim/Bundle/ConnectorBundle/Resources/config/readers.yml (hunk 16 above)
Comment (pierallard) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/readers.yml]: MediaPathTransformer is used by VariantGroupReader too :)

File: CHANGELOG-1.6.md (hunk 21 above)
Comment (pierallard) [inline CHANGELOG-1.6.md]: The 1.5 CsvReader have no construct method, so no arguments!

File: CHANGELOG-1.6.md (hunk 22 above)
Comment (pierallard) [inline CHANGELOG-1.6.md]: The 1.5 CsvProductReader have these arguments:
(
        AttributeRepositoryInterface $attributeRepository,
        array $decimalSeparators,
        array $dateFormats
    )

File: CHANGELOG-1.6.md (hunk 21 above)
Comment (skeleton) [inline CHANGELOG-1.6.md]: ok, sorry, I was checking on master branch :p 
btw, `Pim\Component\Connector\Reader\File\Product\CsvReader` does not exist in 1.5. Good namespace is `Pim\Component\Connector\Reader\File\CsvReader`

File: CHANGELOG-1.6.md [hunk 23]
```diff
@@ -342,3 +333,22 @@
 - Move `Pim\Component\Connector\ArrayConverter\Flat\Product\ColumnsMapper` to `Pim\Component\Connector\ArrayConverter\FlatToStandard\Product\ColumnsMapper`
 - Move `Pim\Component\Connector\ArrayConverter\Flat\Product\AttributeColumnInfoExtractor` to `Pim\Component\Connector\ArrayConverter\FlatToStandard\Product\AttributeColumnInfoExtractor`
 - Move `Pim\Component\Connector\ArrayConverter\Flat\Product\FieldSplitter` to `Pim\Component\Connector\ArrayConverter\FlatToStandard\Product\FieldSplitter`
+- Move `Pim\Component\Connector\Reader\File\Product\CsvReader` to `Pim\Component\Connector\Reader\File\Csv\Reader`. Change constructor to add `Pim\Component\Connector\Reader\File\FileIteratorFactory` and `Pim\Component\Connector\ArrayConverter\ArrayConverterInterface`.
```
Comment (skeleton) [inline CHANGELOG-1.6.md]: Move `Pim\Component\Connector\Reader\File\CsvReader` to `Pim\Component\Connector\Reader\File\Csv\Reader`

File: UPGRADE-1.6.md [hunk 24]
```diff
@@ -50,6 +50,22 @@ Based on a PIM standard installation, execute the following command in your proj
...
+
+The converter services allowed to convert from flat data (e.g. CSV) to standard format. Conversion has moved from processors to readers.
+Now the processors only accept standard format, and the naming has changed from `pim_connector.processor.normalization.<class>.flat` to `pim_connector.processor.normalization.<class>`.
+The YAML Reader moved from BatchBundle to ConnectorBundle; The CSV Readers moved in specific folders, and the naming changed too.
+
+If you use in your import standard Akeneo PIM processor and reader services, please execute the following command in your project folder:
+```
+    find ./src/ -type f -print0 | xargs -0 sed -i 's/pim_connector\.processor\.normalization\.\(.*\)\.flat/pim_connector\.processor\.normalization\.\1/g'
+    find ./src/ -type f -print0 | xargs -0 sed -i 's/pim_base_connector\.reader\.file\.yaml/pim_connector\.reader\.file\.yaml/g'
+    find ./src/ -type f -print0 | xargs -0 sed -i 's/Pim\\Component\\Connector\\Reader\\File\\Product\\CsvReader/Pim\\Component\\Connector\\Reader\\File\\Csv\\CsvReader/g
```
Comment (skeleton) [inline UPGRADE-1.6.md]: find ./src/ -type f -print0 | xargs -0 sed -i 's/Pim\\Component\\Connector\\Reader\\File\\CsvReader/Pim\\Component\\Connector\\Reader\\File\\Csv\\Reader/g

File: UPGRADE-1.6.md [hunk 25]
```diff
@@ -50,6 +50,22 @@ Based on a PIM standard installation, execute the following command in your proj
...
+    find ./src/ -type f -print0 | xargs -0 sed -i 's/Pim\\Component\\Connector\\Reader\\File\\CsvProductReader/Pim\\Component\\Connector\\Reader\\File\\Csv\\CsvProductReader/g
```
Comment (skeleton) [inline UPGRADE-1.6.md]: find ./src/ -type f -print0 | xargs -0 sed -i 's/Pim\\Component\\Connector\\Reader\\File\\CsvProductReader/Pim\\Component\\Connector\\Reader\\File\\Csv\\ProductReader/g

File: src/Pim/Component/Connector/ArrayConverter/FlatToStandard/Product.php [hunk 26]
```diff
@@ -344,7 +329,7 @@ protected function validateItem(array $item, $withRequiredSku)
     protected function validateOptionalFields(array $item)
     {
         $optionalFields = array_merge(
-            ['family', 'enabled', 'categories', 'groups'],
+            ['family', 'enabled', 'categories', 'groups', 'enabled'],
```
Comment (grena) [inline src/Pim/Component/Connector/ArrayConverter/FlatToStandard/Product.php]: :water_buffalo: said:

> Moooh. (Double enabled)

File: src/Pim/Component/Connector/Processor/Denormalization/ProductProcessor.php [hunk 27]
```diff
@@ -177,50 +169,49 @@ protected function filterIdenticalData(ProductInterface $product, array $filtere
      */
     protected function convertItemData(array $item)
     {
-        $this->itemHasStatus = array_key_exists('enabled', $item);
```
Comment (grena) [inline src/Pim/Component/Connector/Processor/Denormalization/ProductProcessor.php]: Empty method?

File: src/Pim/Component/Connector/Reader/File/Csv/Reader.php [hunk 28]
```diff
@@ -80,4 +89,14 @@ public function flush()
     {
         $this->fileIterator = null;
     }
+
+    /**
+     * Returns the options for array converter. It should be overridden in the sub classes.
```
Comment (grena) [inline src/Pim/Component/Connector/Reader/File/Csv/Reader.php]: "it _can_ be", as it's not mandatory :)

File: src/Pim/Component/Connector/Reader/File/Xlsx/Reader.php [hunk 29]
```diff
@@ -75,4 +85,14 @@ public function flush()
     {
         $this->fileIterator = null;
     }
+
+    /**
+     * Returns the options for array converter. It should be overridden in the sub classes.
```
Comment (grena) [inline src/Pim/Component/Connector/Reader/File/Xlsx/Reader.php]: it _can_ be

**Comment by grena** [comment]
Niiiice! GTM once fixed

**Pull request merged**